### PR TITLE
opentelemetry decorator

### PR DESCRIPTION
**Pull request by rebkwok**
Pulling the @instrument decorator from [airlock](https://github.com/opensafely-core/airlock/pull/189) into its own package so we can use it elsewhere.
The last commit does some refactoring and makes some improvements to how it finds attributes from function args.

### CODE REVIEW

File: otel_opensafely/trace.py [hunk 1]
```diff
@@ -38,24 +44,19 @@ def _set_attributes(span, attributes_dict):
         @wraps(func)
         def wrap_with_span(*args, **kwargs):
             name = span_name or func.__qualname__
+            attributes_dict = attributes or {}
+            if func_attributes is not None:
+                func_signature = signature(func)
+                bound_args = func_signature.bind(*args, **kwargs).arguments
```
Comment (bloodearnest) [inline otel_opensafely/trace.py:51]: Huh, TIL. This is really quite nice!

File: otel_opensafely/trace.py [hunk 2]
```diff
@@ -65,3 +66,30 @@ def wrap_with_span(*args, **kwargs):
...
+    4) Finally, raises an exception
+    """
+    if parameter_name in func_kwargs:
+        return func_kwargs[parameter_name]
+    if parameter_name in func_bound_args:
+        return func_bound_args[parameter_name]
+    # an argument for this parameter was not passed explicitly in the function call;
+    # look for a default value
+    param = func_signature.parameters.get(parameter_name)
+    if param and param.default is not Parameter.empty:
```
Comment (bloodearnest) [inline otel_opensafely/trace.py]: really slick!

**Review by bloodearnest — COMMENTED** [review]
The inspect based improvements in that last commit are really nice, great stuff!

However, I'm not sure about creating a separate dependency for this at this stage.  I'd be tempted to let it bake in airlock for a bit before extracting, rather than having an extra dep?

Or are you thinking to release this on pypi as a general open source tool?

**Comment by rebkwok** [comment]
@bloodearnest So, I pulled this out because of your comment about using it in job-server and job-runner too, and I didn't want to copy/paste the decorator code there. But maybe we're best to leave it in just airlock for a while first? I can just update airlock with the changes from the last commit and leave the separate package for now. It's complaining about dependencies on different python versions, and I think job-runner's still on 3.8.

There's nothing opensafely-specific if we were to make it a public package, but it'd probably need a better name.

**Comment by bloodearnest** [comment]
> @bloodearnest So, I pulled this out because of your comment about using it in job-server and job-runner too, and I didn't want to copy/paste the decorator code there.

Ah, sorry, yes, I meant it "at some point"! But the improvements from working on it are great.

> But maybe we're best to leave it in just airlock for a while first? I can just update airlock with the changes from the last commit and leave the separate package for now. It's complaining about dependencies on different python versions, and I think job-runner's still on 3.8.

I just saw and reviewed your PR, nice.

> There's nothing opensafely-specific if we were to make it a public package, but it'd probably need a better name.

Agreed, I think it might be worth doing the work to release to the world, but maybe it can wait until post-initiative deck scrubbing?

**Pull request closed without merging**